File: common/src/main/java/generations/gg/generations/core/generationscore/common/network/packets/C2SEditMailPacket.java
package generations.gg.generations.core.generationscore.common.network.packets;

import generations.gg.generations.core.generationscore.common.GenerationsCore;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;

import java.util.Optional;

public record C2SEditMailPacket(int slot, String contents, Optional<String> title) implements GenerationsNetworkPacket<C2SEditMailPacket> {
    @Override
    public ResourceLocation getId() {
        return ID;
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeVarInt(slot);
        buf.writeUtf(contents, 8192);
        buf.writeOptional(title, (byteBuf, s) -> buf.writeUtf(s, 128));
    }

    public static C2SEditMailPacket decode(FriendlyByteBuf buffer) {
        return new C2SEditMailPacket(buffer.readVarInt(), buffer.readUtf(8192), buffer.readOptional(arg -> arg.readUtf(128)));
    }

    public static ResourceLocation ID = GenerationsCore.id("edit_mail");
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/network/packets/C2SEditMailPacket.kt
package generations.gg.generations.core.generationscore.common.network.packets

import com.cobblemon.mod.common.api.net.NetworkPacket
import generations.gg.generations.core.generationscore.common.GenerationsCore
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.resources.ResourceLocation
import java.util.*

data class C2SEditMailPacket(val slot: Int, val contents: String, val title: Optional<String>) : NetworkPacket<C2SEditMailPacket> {

    override val id: ResourceLocation = ID

    override fun encode(buf: RegistryFriendlyByteBuf) {
        buf.writeVarInt(slot)
        buf.writeUtf(contents, 8192)
        buf.writeOptional(
            title
        ) { byteBuf: FriendlyByteBuf?, s: String? -> buf.writeUtf(s, 128) }
    }

    companion object {
        fun decode(buffer: RegistryFriendlyByteBuf): C2SEditMailPacket {
            return C2SEditMailPacket(
                buffer.readVarInt(),
                buffer.readUtf(8192),
                buffer.readOptional { arg: FriendlyByteBuf ->
                    arg.readUtf(
                        128
                    )
                })
        }

        var ID: ResourceLocation = GenerationsCore.id("edit_mail")
    }
}